in-house

Oracle HSM (originally known as: SAMFS - Storage Archivation Management + QFS)
support of off-laid/staged files. SMB offers "off-line" flag for off-laid files
so the SMB client can behave accordingly to such file-status:
 - explorer do not create thumbnails to avoid staging of all files in browsed dir.
 - explorer display a small "hand-watch" symbol in the corner of file-icon

note: MacOS X ignore the "off-line" flag in SMB protocol.

Originally promoted into the community by:
 https://bugzilla.samba.org/show_bug.cgi?id=5780
...but during the SAMFS version change also the licensing conditions
has changed so the community rejected to adopt the SAMFS support.

diff --git a/source3/modules/wscript_build b/source3/modules/wscript_build
index 09528f3..fca8321 100644
--- a/source3/modules/wscript_build
+++ b/source3/modules/wscript_build
@@ -242,6 +242,14 @@ bld.SAMBA3_MODULE('vfs_solarisacl',
                  internal_module=bld.SAMBA3_IS_STATIC_MODULE('vfs_solarisacl'),
                  enabled=bld.SAMBA3_IS_ENABLED_MODULE('vfs_solarisacl'))
 
+bld.SAMBA3_MODULE('vfs_samfs',
+                 subsystem='vfs',
+                 source='vfs_samfs.c',
+                 deps='samba-util samfs',
+                 init_function='',
+                 internal_module=bld.SAMBA3_IS_STATIC_MODULE('vfs_samfs'),
+                 enabled=bld.SAMBA3_IS_ENABLED_MODULE('vfs_samfs'))
+
 bld.SAMBA3_MODULE('vfs_zfsacl',
                  subsystem='vfs',
                  source='vfs_zfsacl.c',
diff --git a/source3/wscript b/source3/wscript
index 74008a6..48572bb 100644
--- a/source3/wscript
+++ b/source3/wscript
@@ -1755,6 +1755,13 @@ main() {
                       or lack of ceph_statx support''')
             conf.undefine('HAVE_CEPH')
 
+    if conf.CHECK_CFG(package='samfs', args="--cflags --libs", msg='Check for Oracle HSM (aka. SAMFS) API.', uselib_store="SAMFS") :
+        conf.SET_TARGET_TYPE('samfs', 'SYSLIB')
+        conf.CHECK_HEADERS('samfs/lib.h samfs/stat.h', lib='samfs')
+        conf.env['LINKFLAGS_SAMFS'] = [ '-R/opt/SUNWsamfs/lib'+'/'+(conf.CONFIG_GET('LIBPATH_SAMFS')[-1].split('/'))[-1] ]
+        conf.CHECK_FUNCS('sam_stat', lib='samfs')
+        default_shared_modules.extend(TO_LIST('vfs_samfs'))
+
     if Options.options.with_glusterfs:
         conf.CHECK_CFG(package='glusterfs-api', args='"glusterfs-api >= 4" --cflags --libs',
                        msg='Checking for glusterfs-api >= 4', uselib_store="GFAPI")
